client/RpcClient.js

```diff
@@ -0,0 +1,157 @@
+class RpcClient { // eslint-disable-line no-unused-vars
+    /**
+     * @param {Window} targetWindow
+     * @param {string} targetOrigin
+     * @returns {Promise<RpcClientInstance>}
+     * @throws
+     */
+    static async create(targetWindow, targetOrigin) {
+        return new Promise((resolve, reject) => {
+            let connected = false;
+
+            /**
+             * @param {MessageEvent} message
+             */
+            const connectedListener = ({ source, origin, data }) => {
+                if (source !== targetWindow
+                    || data.status !== 'ok'
+                    || data.result !== 'pong'
+                    || data.id !== 1
+                    || (targetOrigin !== '*' && origin !== targetOrigin)) return;
+
+                // Debugging printouts
+                if (data.result.stack) {
+                    const error = new Error(data.result.message);
+                    error.stack = data.result.stack;
+                    console.error(error);
+                }
+
+                window.removeEventListener('message', connectedListener);
+
+                connected = true;
+
+                console.log('RpcClient: Connection established');
+                resolve(new (RpcClient._generateClientClass(targetWindow, targetOrigin))());
+            };
+
+            window.addEventListener('message', connectedListener);
+
+            let connectTimer = 0;
+            const timeoutTimer = setTimeout(() => {
+                window.removeEventListener('message', connectedListener);
+                clearTimeout(connectTimer);
+                reject(new Error('Connection timeout'));
+            }, 10 * 1000);
+
+            /**
+             * Send 'ping' command every second, until cancelled
+             */
+            const tryToConnect = () => {
+                if (connected) {
+                    clearTimeout(timeoutTimer);
+                    return;
+                }
+
+                try {
+                    targetWindow.postMessage({ command: 'ping', id: 1 }, targetOrigin);
+                } catch (e) {
+                    console.error(`postMessage failed: ${e}`);
+                }
+
+                connectTimer = setTimeout(tryToConnect, 1000);
+            };
+
+            connectTimer = setTimeout(tryToConnect, 100);
+        });
+    }
+
+
+    /**
+     * @param {Window} targetWindow
+     * @param {string} targetOrigin
+     * @returns {Newable}
+     * @private
+     */
+    static _generateClientClass(targetWindow, targetOrigin) {
+        /** @type {Newable} */
+        const Client = class {
+            constructor() {
+                // Svub: Code smell that _targetWindow and _waiting are visible outside. TODO later!
+                this._targetWindow = targetWindow;
```

Comment:
there is not really a need for _targetWindow and _targetOrigin and the fix is easy? so why todo later?